Support sandboxed environments

anonhostpi opened the pull request:
- adds `_VERSION` check fix for Luau (the targeted sandbox, which is 5.1 compliant)
- adds stubs for:
  - `io.open`
  - `io.write`

anonhostpi committed:
Support Luau in template environment

anonhostpi committed:
Stub IO when absent

Adds `io.open` and `io.write` stubs when they are absent

bungle commented:
@anonhostpi thanks, interesting.

anonhostpi commented:
Happy to contribute. Will try to reply quickly, if you need any work done

bungle commented:
@anonhostpi, I am wondering what else in this library could be non-sandbox safe. Things like `setfenv`, `load`, and `loadstring` could also be problematic.

anonhostpi commented:
It depends on the sandbox.

The majority of sandboxes are just IO isolated (Lua on WASM, Luau, embedded Lua, etc).

Let me take another gander at it.

anonhostpi commented:
So if we want to address *all sandboxes* (not just IO-sandboxes, but maybe security sandboxes too), we'd need to either annotate minimum requirements and/or make changes to the following

- [x] **load+setfenv (or just loadstring) _is required for the loader to work_ (`jit` is also used here):**
  - *POSSIBLE IMPROVEMENT:* instead of doing a version check (for 5.1), check which function set is available instead
    - _this would eliminate `jit` usage entirely_
  - this is the only time load, setfenv, loadstring, and jit are used.
  - **UPDATE:** _change implemented, needs to be documented on README_
  - **UPDATE 2:** _DONE_
  
https://github.com/bungle/lua-resty-template/blob/2822213d4e0fc0b8b9ac00559a360d099947f002/lib/resty/template.lua#L203-L218

- [x] **_G/_ENV:**
  - this is already partially stubbed:
  - *POSSIBLE IMPROVEMENTS (x2):*
    - use _G and _ENV presence checks instead of a version check
    - add a third optimized option that doesn't include _G/_ENV indexing when they are absent
  - **UPDATE:** *DONE* 

https://github.com/bungle/lua-resty-template/blob/2822213d4e0fc0b8b9ac00559a360d099947f002/lib/resty/template.lua#L68-L77

- [x] **require:**
  - this is only used once, and it is already stubbed out.

- [x] **ngx:**
  - referenced several times, but already stubbed out for non-resty/ngx environments.

- [x] string.dump:
  - not stubbed out, but already calls open, which would throw in IO sandboxes.
  - *POSSIBLE IMPROVEMENT* check that dump and open are defined before declaring
  - **UPDATE:** *change implemented (disabled precompile in applicable sandboxed environments), needs to be documented on README*
  - **UPDATE 2:** *change documented*
    - need to update how I did IO protection, as its currently wrong
  - **UPDATE 3:** *DONE*

https://github.com/bungle/lua-resty-template/blob/2822213d4e0fc0b8b9ac00559a360d099947f002/lib/resty/template.lua#L363-L371

anonhostpi committed:
Adopting type-safety for presence checks

anonhostpi committed:
Fix and improve sandbox checks prior to impl presence checks

anonhostpi committed:
Make sandbox protection explicit

anonhostpi committed:
Replace version checks with presence checks in support of sandboxed envs

anonhostpi committed:
Disable precompile in sandboxed environments

anonhostpi committed:
Clarify compilation and sandboxing details in README

Updated README to clarify compilation and transpilation features, including sandboxing considerations and function descriptions.

anonhostpi commented:
In addition to adding minimum requirements and sandboxing notes to the README (load vs loadstring+setfenv), I also added more technical verbage such as JIT-compiling, AOT-compiling, bytecode (for `template.precompile`), and transpiling (for `template.parse`)

anonhostpi committed:
Allow `template.precompile` in IO sandboxes

anonhostpi committed:
Update `template.precompile` docs

anonhostpi commented:
Files changed tab looks a bit boggling now.

I double-checked, but in case you want to look yourself, there should only be 3 pieces of removed code:

- `jit` - no longer needed as version checks are replaced with presence checks
- `_VERSION` - same
- `path` argument on `transpile.precompile`
  - _still present on the normal 3 `template.precompile...` functions,_ but removed on the 3 IO-sandboxed variations

Only one type of sandbox can't be supported (for obvious reasons) and that is one that doesn't have JIT support (`load`/`loadstring`+`setfenv`). All others should be supportable by the proposed change (IO, `string.dump`, etc)

anonhostpi commented:
I suppose looking back at the comment history of this thread, it might have made sense to add support for the other sandboxes as a secondary PR 😅. Oh well, here they are altogether.
